New structure for the recursive algorithm of the Laplace transform

hanspi42 opened the pull request:
#### References to other Issues or PRs
Part of #24312

#### Brief description of what is fixed or changed
A re-write of the obscure recursive algorithm of the Laplace transform.  The recursion is now much clearer, and it is guaranteed that no infinite recursions can happen through the design of the recursive functions.  More functions, especially with complex arguments, are now covered by the rule set, and several errors have been corrected.

#### Other comments

#### Release Notes
<!-- BEGIN RELEASE NOTES -->
* integrals
  * Improved algorithm for the Laplace transform; `noconds=False` will now always return conditions. Several small errors were corrected, many convergence planes will change, and some results will change as well. May new rules have been added to improve the speed and completeness of computing Laplace transforms.
  * BREAKING CHANGE: The `LaplaceTransform.doit()` method now default to `noconds=True`. This means that calling `doit` returns an expression for the transformed function rather than a tuple with the convergence conditions. This is a bug fix because having `doit` return a tuple breaks using an unevaluated LaplaceTransform as part of an expression (e.g. `(2*LaplaceTransform(exp(t), t, s) - 1).doit()` previously failed because of the tuple). Although this is a bug fix it would likely break any code that currently uses `LaplaceTransform.doit()` directly and depends on the default value of `noconds`. Explicitly setting `noconds` or using the `laplace_transform` function will work the same way as in previous versions of SymPy.
<!-- END RELEASE NOTES -->

hanspi42 committed:
New structure for the recursive algorithm of the Laplace transform

hanspi42 commented:
@oscarbenjamin, @sylee957 and @asmeurer: as two of you suggested, I submit #24312 in multiple parts. Nevertheless, the first change is massive, so my next comment contains a description of all changes I made to the present Laplace transform algorithm.

hanspi42 commented:
This is one part of what I wanted to do in pull request 24312: a re-write of
the obscure recursive algorithm of the Laplace transform.  The recursion is now
much clearer, and it is guaranteed that no infinite recursions can happen through
the design of the recursive functions.

Changes in transforms.py:

1041 ff: The function `_laplace_transform` now gives a bit more debug information

1194 ff: Big change in `_laplace_build_rules`. We used a not so good rule table
(as in public domain and of unknown status by Polyanin) in the past just
because it was publicly available, but since then I found that Bateman's 1954
long tables are in fact publicly available, so I am using those now.  Available:
https://resolver.caltech.edu/CaltechAUTHORS:20140123-101456353 .  Reading
through them also showed me that some conditions were slightly wrong. The whole
table looks totally different now, but the only mathematical changes are
changes in the conditions for many rules that extend their validity in the
complex domain. Polyanin's tables are restricted to the real domain. The
conditions have changed in the rules that are now numbered 4.2.6, .18, .22,
.20, 4.3.4, .7, 4.5.3, .21, .22, .25, .26, .27, .28, .29, .31, .33, 4.6.4, .5,
.11, 4.12.2, .6, .11, 4.14.1, .7, .8, .30, 4.15.19, 4.16.1, .6, .7, .18  .  In
case you want to look at Bateman's tables, please be aware of one pitfall:
he uses the symbol `gamma` to denote what we would call `exp(EulerGamma)`; for
him, `EulerGamma` is called `C`, the Euler-Mascheroni constant.

1394 ff: `_laplace_cr` has been changed and a new function `_laplace_ct` has been
added because all other functions now always pass tuples around, no matter
whether `noconds` is True or False.

From here on all rules have been rewritten such that they either return a
tuple `True, (F, p, c)` when they could do something, or `False, (f, p, c)`
when they could not. If they can modify an incoming function, they may call
`LaplaceTransform` again. All of the functions are written such that they will
never call `LaplaceTransform` with a trivial modification that would cause an
infinite recursion. All functions immediately convert incoming non-tuples into
tuples by applying `_laplace_ct`. Some of the functions also passed the hint
`simplify`; this does not make sense anymore with the new top-level structure.
Finally, the docs for the helper functions have been improved.

1416 ff: `_laplace_rule_timescale` covered `a==1` in the past, but it must not
do that because that could cause an infinite recursion. This is a bug fix.

1440 ff: `_laplace_rule_heaviside` only worked for positive time shifts. Now it
also covers negative time shifts by simply removing the respective `Heaviside`
function as it is always `1` in the integration range of the LT.

1477 ff: `_laplace_rule_exp` made the plane correction incorrectly, it added
`a` instead of `re(a)`.

1501 ff: `_laplace_rule_delta` is new. It evaluates products with `DiracDelta`
by applying the masking property, but only if the impulse is on the positive
real axis, otherwise it will return `0`.

1532 ff: `_laplace_rule_trig` now adapts the planes correctly even if the
arguments of the trigonometric functions are complex.

1576 ff: `_laplace_rule_diff` does the same as before.

1605 ff: `_laplace_rule_sdiff` is new, it finds some `p(t)*f(t)` where `p(t)`
is a polynomial in `t` and evaluates rule 4.1.6 in an efficient way by taking
the LT of `f(t)` once and calculating all required frequency derivatives.

1644 ff: `_laplace_expand` is new. It attempts to expand the incoming function
with progressively stronger expansion algorithms and, if it can split the
expreession into an `Add`, takes the LT of that new expression. In order to
prevent infinite recursions, it will refuse to do anything if the incoming
function is `Add`. The idea behind this scheme is that many of the simple
rules will not trigger if the expansion is made too soon. So the levels of
expansions are: expand on top level, expand multiplications, expand, and
finally even expand trigonometric functions.  The rationale behind this
layered expansion is that the simple rules will always be checked before
`_laplace_expand` is called again, so in a way it is optimized to use as many
simple rules as possible.

1674 ff: The original `_laplace_apply_rules` applied both simple and program
rules, it is now split in two parts, `_laplace_apply_prog_rules` and
`_laplace_apply_simple_rules`.

1726 ff: Many problems with the exisiting implementation occurred because
the class `LaplaceTransform` attempted to re-use the `doit` inherited
from the class `IntegralTransform`. This is now not the case anymore, which
makes everything easier to read, but changes a lot in the class. In addition
to this, it was not certain that using `.doit(noconds=False)` would return
a tuple; this is now always the case.

The changes within the class in detail:

1742 ff: `_compute_transform` does not apply rules anymore; `doit` does
this now.

1763 ff: `_try_directly` does not attempt expansion anymore; `doit` does
this now.

1776 ff: `LaplaceTransform` now has its own `doit`. The rationale behind the
implementation is that it looks at additive terms independently and, for each
sum term, attempts to apply simple rules, apply program rules, expand the
incoming function, and, if everything else fails, tries to compute the
transform.  The program rules and the expander will call `doit` again, and
all functions always exchange tuples, so `doit` collects all results in `r`,
all convergence planes in `p` and all conditions in `c`. At the very beginning,
it removes any factor `Heaviside(t)` unless there is a `DiracDelta(t)` in the
rest, because `Heaviside(t)*DiracDelta(t)` would be `DiracDelta(t)/2`, which
will later be handled by the program rules.

1843 ff: `laplace_transform` also used to sometimes return just a function
even if `noconds=False`; this is now not the case anymore. The only other change
here is an updated doc string to refer to the rule-base we use, 


These are the changes in transforms.py, but this also causes many changes to
test_transforms.py and test_wester.py. The latter are only necessary because
the order of the terms in the output change, but the former are less trivial.
There are many changes and additions to `test_laplace_transform`; I have
reviewed all of them myself, but it would not be bad if someone else had
a look at it. `test_as_integral` requires a small change because, without
`noconds=True`, conditions will be returned. `test_issue_14692` has been
removed; this actually tested the behaviour of Meijer integration through
a specific Laplace transform, but the rule set here is extended to complex
arguments and covers the function used in `test_issue_14692`, so Issue
#14692 cannort be tested through the LT anymore.

ThePauliPrinciple commented:
Not very familiar with Laplace transforms, but some of the conditions appear to have changed (e.g. 660:`a`->653:`-a`). I'm assuming the new results are correct/better. If indeed the new conditions are mathematically different, then the release notes should state that this is the case.

sylee957 commented:
> From here on all rules have been rewritten such that they either return a
tuple True, (F, p, c) when they could do something, or False, (f, p, c)
when they could not.

The tuple with status value may be refactored to use [optional monad](https://docs.python.org/3/library/typing.html#typing.Optional) (`tuple[..., ..., ...] | None`) 
but that is my preference.

hanspi42 commented on sympy/integrals/transforms.py:
```diff
@@ -1186,499 +1194,363 @@ def _laplace_build_rules(t, s):
-     (log(s+sqrt(s**2-a**2)))/(sqrt(s**2-a**2)),
-     a>0, Abs(a), dco)
+    (besselk(0, a*t), (log(s+sqrt(s**2-a**2)))/(sqrt(s**2-a**2)),
+     S.true, -re(a), dco) # 4.16.23
```
This is incorrect, the result should be `(log( (s + sqrt(s**2-a**2))/a )/(sqrt(s**2-a**2))` according to Bateman54, p. 197.  The convergence plane is OK, though.

hanspi42 replied:
Change (F) explained in comment below.

hanspi42 commented:
> but some of the conditions appear to have changed (e.g. 660:`a`->653:`-a`). I'm assuming the new results are correct/better.

They are, but while re-checking whether this one is really more correct/better, I saw that the old one is actually wrong. It seems that nobody noticed; maybe nobody every needs the Laplace transform of `besselk(0, a*t)` ;)

hanspi42 commented on sympy/integrals/transforms.py:
```diff
@@ -1186,499 +1194,363 @@ def _laplace_build_rules(t, s):
+      sd: shift direction; shift along real or imaginary axis if `1` or `I`
+
+    The convergence plane is changed only if the frequency shift is done along
+    the real axis.2
```
Better: `if a frequency shift ...`; there is also a spurious `2` at the end of `axis.2`

hanspi42 commented:
@oscarbenjamin, @sylee957 and @asmeurer: This is a bit anticlimactic here, so I ask: did nobody have time to look at it until now, or is it already good enough if I correct the small errors outlined above? I don't want to hurry anyone, but I also don't want to wait needlessly when nothing is going to happen ;)

sylee957 commented:
I'm generally okay with the new code. But I'd defer the merging until other review approvals are done.
And I don't think that we need to correct the mathematical accuracy at the point now because the source is clear.
But I think that some type analysis about the code can be done in the future, because there are many pragmatic changes in the code.

hanspi42 commented:
> I'm generally okay with the new code. But I'd defer the merging until other review approvals are done.

Does someone else already have the task to do one? I don't see this.

> And I don't think that we need to correct the mathematical accuracy at the point now because the source is clear.

So we'd first merge this and then make a very small additional pull request?

hanspi42 commented:
> But I think that some type analysis about the code can be done in the future, because there are many pragmatic changes in the code.

Good idea, I think, is this about using monads again? That'd mean introducing one abstraction layer more in the code, wouldn't it? If so, then I think I should first do the inverse tranasform after this PR, and only when both are merged do the type analysis.

hanspi42 committed:
Corrected a small error in the new code and added a test case for it

oscarbenjamin commented:
This gives a TypeError:
```python
In [2]: laplace_transform(Heaviside(t - y)*g(t), t, s)
---------------------------------------------------------------------------
TypeError
```

oscarbenjamin commented on sympy/integrals/transforms.py:
```diff
@@ -1186,499 +1194,363 @@ def _laplace_build_rules(t, s):
+    """
+    # This rule is not in Bateman54
+    fn, p, c = _laplace_ct(f)
+    k, f = fn.as_independent(t, as_Add=False)
```
Lots of rules apply `as_independent` as their first step. Could this not be handled higher up?

hanspi42 replied:
Easily. I didn't even notice that it is now almost everywhere. Change (B) in my list below.

oscarbenjamin commented on sympy/integrals/transforms.py:
```diff
@@ -1186,499 +1194,363 @@ def _laplace_build_rules(t, s):
+            return True, (k*r, Max(p+re(ma2[a]), pr), And(c, cr))
+    return False, (fn, p, c)
+
+def _laplace_rule_delta(f, t, s):
```
Can this not be a simple rule?

hanspi42 replied:
As it is, it can, but later in the development I am also planning to cover mor complicated arguments inside `DiracDelta`, and my instints tell me to keep it separate from the simple rules.

oscarbenjamin commented on sympy/integrals/transforms.py:
```diff
@@ -1826,19 +1760,84 @@ def _collapse_extra(self, extra):
+                results.append(r)
+                continue
+            T = None
+            try_directly = not any(func.has(t_)
+                                   for func in f.atoms(AppliedUndef))
+            k_, f_ = f.as_independent(t_, as_Add=False)
+            if try_directly:
+                try:
+                    T = self._compute_transform(f_, t_, s_)
+                except IntegralTransformError:
+                    T = None
+            if T is not None:
+                r_, p_, c_ = _laplace_ct(T)
+                results.append((k_*r_, p_, c_))
+            else:
+                results.append(_laplace_ct(k_*LaplaceTransform(f_, t_, s_)))
```
Perhaps make this a separate function like the others.

hanspi42 replied:
I'd rather not, afetr removing the `as_independent`, this is about very few lines.

oscarbenjamin commented on sympy/integrals/transforms.py:
```diff
@@ -1826,19 +1760,84 @@ def _collapse_extra(self, extra):
-        return fn, LT
 
+        if fn.has(Heaviside(t_)) and not fn.has(DiracDelta(t_)):
+            fn = fn.replace(Heaviside(t_), 1)
```
Is this not handled by the Heaviside rule?

hanspi42 replied:
Yes, it is. Change (C) in my list below.

oscarbenjamin commented:
The structure here as I understand it is more or less like this:
```python
def laplace_transform(expr):
    simple_rules = build_rules()
    newterms = []
    for term in Add.make_args(expr):
        for rule in simple_rules:
             if matches(rule, term):
                 newterms.append(apply(rule, term))
                 continue
        for rule in prog_rules:
            if matches(rule, term):
                 newterms.append(apply(rule, laplace_transform(part(term)))))
                 continue
        for rule in expand_rules:
            if apply(rule, expr) != expr:
                newterms.append(laplace_transform(apply(rule, expr)))
        newterms.append(do_integral(expr))
    return simplify(Add(*newterms))
```
I'm trying to understand this from a computational performance perspective. We have a large number of simple rules. Every recursion results in us trying them all again. If several `prog_rules` apply then we will have to several times go through all the `simple_rules` before getting to the point where they apply.

Note in the timings below that SymPy uses a cache for many operations so often running an operation the second time is faster than the first. I'm mostly showing results that are the second time so that various things are already cached (e.g. the simple rules).

Running this under the profiler in isympy shows that the slowest thing in simple cases is the call to `_laplace_build_rules`:
```python
In [11]: %prun -s cumulative laplace_transform(exp(t), t, s)

         65144 function calls (61745 primitive calls) in 0.086 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.086    0.086 {built-in method builtins.exec}
        1    0.000    0.000    0.086    0.086 <string>:1(<module>)
        1    0.000    0.000    0.085    0.085 transforms.py:1842(laplace_transform)
        1    0.000    0.000    0.085    0.085 transforms.py:1775(doit)
        1    0.000    0.000    0.072    0.072 transforms.py:1692(_laplace_apply_simple_rules)
        1    0.002    0.002    0.059    0.059 transforms.py:1193(_laplace_build_rules)
...
```
In more complicated cases this gets called more than once:
```python
%prun -s cumulative laplace_transform(exp(t)*sin(t)*t**2, t, s)

         892498 function calls (826390 primitive calls) in 0.756 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.757    0.757 {built-in method builtins.exec}
        1    0.000    0.000    0.756    0.756 <string>:1(<module>)
        1    0.000    0.000    0.752    0.752 transforms.py:1842(laplace_transform)
      3/1    0.000    0.000    0.752    0.752 transforms.py:1775(doit)
      2/1    0.000    0.000    0.473    0.473 transforms.py:1673(_laplace_apply_prog_rules)
      2/1    0.000    0.000    0.462    0.462 transforms.py:1476(_laplace_rule_exp)
        3    0.000    0.000    0.436    0.145 transforms.py:1692(_laplace_apply_simple_rules)
        3    0.000    0.000    0.281    0.094 basic.py:1814(simplify)
        3    0.000    0.000    0.281    0.094 simplify.py:420(simplify)
26853/22440    0.038    0.000    0.280    0.000 cache.py:69(wrapper)
        3    0.005    0.002    0.256    0.085 transforms.py:1193(_laplace_build_rules)
```
We can also see that lots of time is spent in simplify which is called several times even though its only job is to simplify the return value. It gets called for each level of recursion.

I'm wondering if it makes sense to make some changes:

- Maybe the rule building should be cached. It shouldn't be necessary to build the rules more than once.
- Maybe `prog_rules` should be applied before simple rules and the structure should be such that we don't recurse back to trying the same `prog_rules` that were already applied (I think these rules are all idempotent).
- There are many places that use `as_independent`. Maybe that should just be the first step of `laplace_transform`
- There should be some way to avoid calling simplify and probably it should not be used by default (simplifying on output is especially problematic because then it would be easy for the user to call simplify themselves).

I'm not sure exactly what the different levels of expansion are for but they potentially mean rerunning a lot of the simple rules and prog rules many times because each time some expansion is possible we recurse back to top level. Likewise the more prog rules apply to an expression the slower it gets because we have to loop through everything several times.

This gives an error:
```python
laplace_transform((exp(t)*sin(t)*t**2*Heaviside(t-3)-1)*exp(-t), t, s)
```
Instead if I try this it hangs:
```python
laplace_transform((exp(t)*sin(t)*t**2*Heaviside(t-3)-1)*sin(-t), t, s)
```
That's because it's calling simplify with a big expression. There doesn't seem to be a way to disable simplify.

This is slow:
```python
In [13]: %prun -s cumulative laplace_transform((exp(t)*sin(t)*t**2*Heaviside(t-3)-1)*t, t, s)

         17428672 function calls (15871915 primitive calls) in 12.690 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000   12.696   12.696 {built-in method builtins.exec}
        1    0.000    0.000   12.696   12.696 <string>:1(<module>)
        1    0.000    0.000   12.673   12.673 transforms.py:1842(laplace_transform)
      7/1    0.000    0.000   12.673   12.673 transforms.py:1775(doit)
        7    0.000    0.000   11.347    1.621 basic.py:1814(simplify)
     84/7    0.001    0.000   11.347    1.621 simplify.py:420(simplify)
 47125/35    0.055    0.000    7.587    0.217 traversal.py:228(bottom_up)
      6/1    0.000    0.000    7.387    7.387 transforms.py:1673(_laplace_apply_prog_rules)
      4/1    0.000    0.000    7.359    7.359 transforms.py:1604(_laplace_rule_sdiff)
        7    0.000    0.000    6.711    0.959 trigsimp.py:463(trigsimp)
        7    0.000    0.000    6.711    0.959 trigsimp.py:557(<lambda>)
        7    0.000    0.000    6.711    0.959 trigsimp.py:1127(futrig)
      257    0.002    0.000    6.706    0.026 trigsimp.py:1170(_futrig)
```
Note that `simplify` is called 7 times most likely with more or less the same input each time.

hanspi42 commented:
> This gives an error:
> 
> ```python
> laplace_transform((exp(t)*sin(t)*t**2*Heaviside(t-3)-1)*exp(-t), t, s)
> ```
> 
> Instead if I try this it hangs:
> 
> ```python
> laplace_transform((exp(t)*sin(t)*t**2*Heaviside(t-3)-1)*sin(-t), t, s)
> ```

This is the same thing twice ;)

hanspi42 committed:
Several small changes due to reviewers' comments

hanspi42 commented:
Thankls everybody for the reviews. Here are the changes I made; I will also refer to them in the answers to the comments.

Changes after reviews:

(A) In `_laplace_rule_timescale` and `_laplace_rule_heaviside`, a `TypeError` is
prevented by inserting a check for `ma2[a].is_real` before the check for
`ma2[a]>0`. This makes sure that things like `laplace_transform(Heaviside(t-y)*g(t), t, s)`
do not give a type error anymore when `y` is not real.

(B) Moved `.as_independent(t_, as_Add=False)` up into `doit`.  It is now removed
from `_laplace_rule_timescale, _laplace_rule_heaviside, _laplace_rule_exp,
_laplace_rule_delta, _laplace_rule_trig, _laplace_apply_simple_rules` and
one time more inside `doit`.

(C) Check for `Heaviside(t)` removed from `doit`.

(D) Changed `_laplace_rule_sdiff` such that it will not crash when it has
`s+1` as its input.

(E) Made it possible to switch simplification off. `doit` now actually reads
the `simplify` hint, and all of the rules now call `LaplaceTransform` with
`simplify=False` inside the recursion, these are: `_laplace_rule_timescale,
_laplace_rule_heviside, _laplace_rule_exp, _laplace_rule_trig,
_laplace_rule_diff, _laplace_rule_sdiff, _laplace_expand`. This has uncovered
a convergence plane error in `_laplace_rule_exp` and in the simple and program
rules for `DiracDelta`, which in turn has changed a few convergence planes in
the test cases. I have verified all new planes by calculating the poles of
the results.

(F) Corrected the incorrect simple rule for `besselk`.

oscarbenjamin commented:
> a `TypeError` is
> prevented by inserting a check for `ma2[a].is_real` before the check for
> `ma2[a]>0`

The property to check is `is_comparable`. What `is_comparable` means is that the expression can be numerically evaluated to a real float.

oscarbenjamin commented:
> > Instead if I try this it hangs:
> > That's because it's calling simplify with a big expression. There doesn't seem to be a way to disable simplify.
> 
> Simplification can now be switched off by `simplify=False`

But it is still used by default. This example does not need to be slow:
```
laplace_transform((exp(t)*sin(t)*t**2*Heaviside(t-3)-1)*sin(-t), t, s)
```
This is precisely the kind of thing that this rule-based engine should be able to handle efficiently. I think part of why simplify is slow is that a complicated expression is constructed.

oscarbenjamin commented:
> > * Maybe the rule building should be cached. It shouldn't be necessary to build the rules more than once.
> 
> At least not if the incoming `s` and `t` are the same. I'd like to look at this later, it is non-trivial.

I think it should be fairly simple if we just replace the symbols `s` and `t` in the input expression with dummies like:
```
In [10]: _s, _t = symbols('s, t', cls=Dummy)

In [11]: (t*(cos(t) + sin(t))).xreplace({t:_t})
Out[11]: t⋅(sin(t) + cos(t))
```
Then the rules can be built once with `_s` and `_t` as the symbols.

oscarbenjamin commented:
I think this example should be a lot faster:
```python
In [23]: %prun -s cumulative laplace_transform((t**10 + t**5 + 1)*exp(t), t, s)
         964566 function calls (875257 primitive calls) in 0.767 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.768    0.768 {built-in method builtins.exec}
        1    0.000    0.000    0.768    0.768 <string>:1(<module>)
        1    0.000    0.000    0.766    0.766 transforms.py:1855(laplace_transform)
      2/1    0.000    0.000    0.766    0.766 transforms.py:1782(doit)
        4    0.000    0.000    0.492    0.123 transforms.py:1700(_laplace_apply_simple_rules)
27962/21337    0.039    0.000    0.385    0.000 cache.py:69(wrapper)
        4    0.005    0.001    0.382    0.096 transforms.py:1193(_laplace_build_rules)
5247/4634    0.007    0.000    0.363    0.000 decorators.py:224(_func)
4944/4519    0.006    0.000    0.329    0.000 decorators.py:99(binary_op_wrapper)
 1031/714    0.014    0.000    0.319    0.000 operations.py:52(__new__)
  645/471    0.023    0.000    0.280    0.001 mul.py:197(flatten)
        1    0.000    0.000    0.260    0.260 basic.py:1814(simplify)
        1    0.000    0.000    0.260    0.260 simplify.py:420(simplify)
        1    0.000    0.000    0.242    0.242 transforms.py:1681(_laplace_apply_prog_rules)
        1    0.000    0.000    0.235    0.235 transforms.py:1475(_laplace_rule_exp)
...
```
The simple rules are applied 4 times. It should just be one application of polynomial rule and one of exponential.

oscarbenjamin commented:
> > Simplification can now be switched off by `simplify=False`
> 
> But it is still used by default.

I tried disabling the call to simplify and running the test suite and for the most part it does not significantly alter the return values. The main effect of `simplify` is really just that it calls `cancel`. Since `cancel` ca be one of the slower parts of `simplify` it would still be good to avoid that.

In many cases the only difference is really what `signsimp` would do. Note that calling `signsimp` is fine: it is never a slow function.

The rules where `simplify` makes a big difference are those involving `sin`, `cos`, `sinh` and `cosh` with either shifts or powers:
```python
In [12]: print(laplace_transform(sin(t)**3, t, s, simplify=False)[0])
-I*(-2/((s + I)*((s + I)**2 + 4)) + 2/((s - I)*((s - I)**2 + 4)))/2

In [13]: print(laplace_transform(sin(t)**4, t, s, simplify=False)[0])
-I*(-I*(2/((s - 2*I)*((s - 2*I)**2 + 4)) - 2/(s*(s**2 + 4)))/2 + I*(-2/((s + 2*I)*((s + 2*I)**2 + 4)) + 2/(s*(s**2 + 4)))/2)/2

In [14]: print(laplace_transform(sin(t)**5, t, s, simplify=False)[0])
-I*(-I*(-I*(-2/((s - I)*((s - I)**2 + 4)) + 2/((s - 3*I)*((s - 3*I)**2 + 4)))/2 + I*(-2/((s + I)*((s + I)**2 + 4)) + 2/((s - I)*((s - I)**2 + 4)))/2)/2 + I*(-I*(-2/((s + I)*((s + I)**2 + 4)) + 2/((s - I)*((s - I)**2 + 4)))/2 + I*(-2/((s + 3*I)*((s + 3*I)**2 + 4)) + 2/((s + I)*((s + I)**2 + 4)))/2)/2)/2

In [15]: print(laplace_transform(sin(t)**6, t, s, simplify=False)[0])
-I*(-I*(-I*(-I*(-2/((s - 2*I)*((s - 2*I)**2 + 4)) + 2/((s - 4*I)*((s - 4*I)**2 + 4)))/2 + I*(2/((s - 2*I)*((s - 2*I)**2 + 4)) - 2/(s*(s**2 + 4)))/2)/2 + I*(-I*(2/((s - 2*I)*((s - 2*I)**2 + 4)) - 2/(s*(s**2 + 4)))/2 + I*(-2/((s + 2*I)*((s + 2*I)**2 + 4)) + 2/(s*(s**2 + 4)))/2)/2)/2 + I*(-I*(-I*(2/((s - 2*I)*((s - 2*I)**2 + 4)) - 2/(s*(s**2 + 4)))/2 + I*(-2/((s + 2*I)*((s + 2*I)**2 + 4)) + 2/(s*(s**2 + 4)))/2)/2 + I*(-I*(-2/((s + 2*I)*((s + 2*I)**2 + 4)) + 2/(s*(s**2 + 4)))/2 + I*(-2/((s + 4*I)*((s + 4*I)**2 + 4)) + 2/((s + 2*I)*((s + 2*I)**2 + 4)))/2)/2)/2)/2
```
What you can see is that the rules for this systematically make very large expressions which simplify then struggles to bring back down:
```python
In [17]: %time F.simplify()
CPU times: user 1.02 s, sys: 0 ns, total: 1.02 s
Wall time: 1.02 s
Out[17]: 
             720              
──────────────────────────────
  ⎛ 6       4        2       ⎞
s⋅⎝s  + 56⋅s  + 784⋅s  + 2304⎠
```
A similar thing happens with shifts
```python
In [21]: laplace_transform(sin(t-x), t, s, simplify=False)[0]
Out[21]: 
    ⎛  cos(x)   cos(x)⎞                        
  ⅈ⋅⎜- ────── + ──────⎟                        
    ⎝  s + ⅈ    s - ⅈ ⎠     sin(x)      sin(x) 
- ───────────────────── - ───────── - ─────────
            2             2⋅(s + ⅈ)   2⋅(s - ⅈ)

In [22]: simplify(_)
Out[22]: 
-s⋅sin(x) + cos(x)
──────────────────
       2          
      s  + 1 
```
This one is far too slow:
```python
In [28]: %time laplace_transform(sin(t-2)*cos(t-2), t, s, simplify=False)[0]
CPU times: user 4.63 s, sys: 0 ns, total: 4.63 s
Wall time: 4.63 s
Out[28]: 
    ⎛             2                 2   ⎞     ⎛           2                 2   ⎞                                                              
    ⎜  (s - ⅈ)⋅sin (2)   (s + ⅈ)⋅sin (2)⎟     ⎜(s - ⅈ)⋅cos (2)   (s + ⅈ)⋅cos (2)⎟                                                              
  ⅈ⋅⎜- ─────────────── + ───────────────⎟   ⅈ⋅⎜─────────────── - ───────────────⎟                                                              
    ⎜           2                 2     ⎟     ⎜         2                 2     ⎟                                                              
    ⎝    (s - ⅈ)  + 1      (s + ⅈ)  + 1 ⎠     ⎝  (s - ⅈ)  + 1      (s + ⅈ)  + 1 ⎠                   ⎛                  2         ⎞     sin(4)  
- ─────────────────────────────────────── - ───────────────────────────────────── + LaplaceTransform⎝-sin(2)⋅cos(2)⋅cos (t), t, s⎠ + ──────────
                     2                                        2                                                                        ⎛ 2    ⎞
                                                                                                                                     s⋅⎝s  + 4⎠
```
It also hasn't even finished leaving an easily manageable left over LaplaceTransform:
```python
In [29]: %time _.doit()
CPU times: user 164 ms, sys: 4 ms, total: 168 ms
Wall time: 168 ms
Out[29]: 
    ⎛             2                 2   ⎞     ⎛           2                 2   ⎞                               
    ⎜  (s - ⅈ)⋅sin (2)   (s + ⅈ)⋅sin (2)⎟     ⎜(s - ⅈ)⋅cos (2)   (s + ⅈ)⋅cos (2)⎟                               
  ⅈ⋅⎜- ─────────────── + ───────────────⎟   ⅈ⋅⎜─────────────── - ───────────────⎟                               
    ⎜           2                 2     ⎟     ⎜         2                 2     ⎟   ⎛ 2    ⎞                    
    ⎝    (s - ⅈ)  + 1      (s + ⅈ)  + 1 ⎠     ⎝  (s - ⅈ)  + 1      (s + ⅈ)  + 1 ⎠   ⎝s  + 2⎠⋅sin(4)     sin(4)  
- ─────────────────────────────────────── - ───────────────────────────────────── - ─────────────── + ──────────
                     2                                        2                           ⎛ 2    ⎞      ⎛ 2    ⎞
                                                                                      2⋅s⋅⎝s  + 4⎠    s⋅⎝s  + 4⎠

In [30]: %time _.simplify()
CPU times: user 456 ms, sys: 0 ns, total: 456 ms
Wall time: 455 ms
Out[30]: 
  s⋅sin(4)         
- ──────── + cos(4)
     2             
───────────────────
        2          
       s  + 4 
```

oscarbenjamin commented on sympy/integrals/transforms.py:
```diff
@@ -1186,499 +1194,360 @@ def _laplace_build_rules(t, s):
-    #   s2: sign of the second copy of the Laplace transform of z
-    #   sd: shift direction; shift along real or imaginary axis if `1` or `I`
     trigrules = [(sinh(y), '1.6',  1, -1, 1), (cosh(y), '1.7', 1, 1, 1),
                  (sin(y),  '1.8', -I, -1, I), (cos(y),  '1.9', 1, 1, I)]
```
In a power like `sin(t)**10 * f(t)` this will extract only one factor of `sin(t)` at a time which in efficient. It should be possible to derive the more general rule for powers e.g.:
```python
In [23]: st = lambda f: (-I*f.subs(s, s-I) - f.subs(s, s+I))/2

In [24]: F = Function('F')(s)

In [25]: st(F)
Out[25]: 
  ⅈ⋅F(s - ⅈ)   F(s + ⅈ)
- ────────── - ────────
      2           2    

In [26]: st(st(F))
Out[26]: 
    ⎛  F(s)   ⅈ⋅F(s - 2⋅ⅈ)⎞                      
  ⅈ⋅⎜- ──── - ────────────⎟                      
    ⎝   2          2      ⎠   ⅈ⋅F(s)   F(s + 2⋅ⅈ)
- ───────────────────────── + ────── + ──────────
              2                 4          4     

In [27]: st(st(F)).simplify()
Out[27]: 
ⅈ⋅F(s)   F(s - 2⋅ⅈ)   F(s + 2⋅ⅈ)
────── - ────────── + ──────────
  2          4            4     

In [28]: st(st(st(F))).simplify()
Out[28]: 
ⅈ⋅F(s - 3⋅ⅈ)   3⋅F(s - ⅈ)   3⋅ⅈ⋅F(s + ⅈ)   F(s + 3⋅ⅈ)
──────────── + ────────── - ──────────── - ──────────
     8             8             8             8  
```
The coefficients follow a pattern with binomial coefficients:
```
In [39]: factor_terms(expand((-(x + I)/2)**3))
Out[39]: 
 3        2          
x  + 3⋅ⅈ⋅x  - 3⋅x - ⅈ
─────────────────────
          8 
```
The function `st` generates a complexified version of the Pascal's triangle. For `cosh` it actually is the triangle.

oscarbenjamin replied:
Also why doesn't the second example work:
```
In [86]: laplace_transform(cosh(t)*f(t), t, s)
Out[86]: 
⎛LaplaceTransform(f(t), t, s - 1)   LaplaceTransform(f(t), t, s + 1)          ⎞
⎜──────────────────────────────── + ────────────────────────────────, -∞, True⎟
⎝               2                                  2                          ⎠

In [87]: laplace_transform(cosh(t)**2*f(t), t, s)
Out[87]: 
⎛                ⎛         2         ⎞          ⎞
⎝LaplaceTransform⎝f(t)⋅cosh (t), t, s⎠, -∞, True⎠

```

hanspi42 commented:
Oh well, I made a stupid mistake, sorry. I moved the `.as_independent` to the wrong place, it does now not act on the terms of `Add` anymore, just on the whole sum, so the recursive rule set now works in a different and much less efficient way. I did not notice this because the algorithm is so robust that the mistake did not cause more test cases to be solved by integration.

I will solve this in the coming days and reply to all your comments in one go.

hanspi42 commented:
> > a `TypeError` is
> > prevented by inserting a check for `ma2[a].is_real` before the check for
> > `ma2[a]>0`
> 
> The property to check is `is_comparable`. What `is_comparable` means is that the expression can be numerically evaluated to a real float.

That will not help.

```
a = symbols('a', positive=True)
x = a
x = a+1
x = 1/a
x = a-1
```

All of the above give `False` under `.is_comparable` but the correct truth value with `x>0`

In addition, I don't want the rules to trigger for complex arguments.

oscarbenjamin commented:
There is a function `is_gt`:
```python
In [19]: from sympy.core.relational import is_gt

In [20]: a = symbols('a', positive=True)

In [21]: is_gt(a, S(0))
Out[21]: True

In [22]: is_gt(a+1, S(0))
Out[22]: True

In [23]: is_gt(1/a, S(0))
Out[23]: True

In [24]: is_gt(a-1, S(0)) # None because unknown

In [25]: is_gt(sqrt(-1), S(0)) # doesn't raise

In [26]: is_gt(-a, S(0))
Out[26]: False
```

hanspi42 commented:
> ```python
> 
> In [24]: is_gt(a-1, S(0)) # None because unknown
> ```

Exactly. That will not work inside an `if`: `TypeError: cannot determine truth value of Relational`

Is there a way to modify an expression such that `True -> True` and `all else -> False`?

oscarbenjamin commented:
> ```python
> In [24]: is_gt(a-1, S(0)) # None because unknown
> ```
> 
> Exactly. That will not work inside an if: TypeError: cannot determine truth value of Relational

Have you tried it?

```python
In [27]: if is_gt(a-1, S(0)):
    ...:     print("True")
    ...: else:
    ...:     print("None or False")
    ...: 
None or False
```

hanspi42 commented:
> Have you tried it?

Yes and no. I stupidly thought that `Gt` does the same. Thanks.

oscarbenjamin commented:
Btw if you're comparing against zero then you can just do `expr.is_positive` rather than `is_gt(expr, S.Zero)`. These should be equivalent and both use three valued logic True, False or None. Actually the equivalent property is `expr.is_extended_positive` which also allows for `oo`. Whether `oo` should be considered permissible is something that has to be considered in context.

hanspi42 commented:
Very good. By the way, shall I change the default to `simplify=False`?

hanspi42 committed:
Changes after performance reviews

hanspi42 commented:
> > shall I change the default to `simplify=False`
> 
> Yes, I think so.

There you go:

Changes after performance reviews:

--

(G) Corrected change (B); `.as_independent(t_, as_Add=False)` also needs to be
    applied to the individual sum terms.

(H) Replaced checks `>0` and `<0` by `.is_positive` and `.is_negative` in
    `_laplace_rule_timescale` and `_laplace_rule_heaviside`.
    
(I) Changed the default to `simplify=False` and adapted the relevant test
    cases, mostly by adding `simplify=True` to the test case such that the
    expected result is easier to read and verify for humans and also to
    make it easier to see that nothing mathematical changed.

--

I see that multiple powers of trigonometric functions could be handled much
more efficiently, but since this part is an existing implementation, I will
change it in a later PR in order to not mix it up with the change of the
algorithm structure.

My plan is to make a list of all ideas to improve the LT, but to switch to
the ILT first because one of the most important issues is the huge 
discrepancy between LT and ILT in both speed and capabilities.  I'll 
switch bacck to the LT later.

hanspi42 commented:
P.S. The change (G) above will solve several of the speed and complicated-result problems you saw.

hanspi42 committed:
Removed un-used import.

hanspi42 committed:
Small correction in one test

oscarbenjamin commented:
Is this correct?
```python
In [55]: F, a, cond = laplace_transform(cosh(t)**2, t, s)

In [56]: F.factor()
Out[56]: 
   2      
  s  - 2  
──────────
 2        
s ⋅(s - 4)

In [57]: a
Out[57]: 2
```
I thought this means it should converge for `s > 2` but there is a pole at `s = 4`.

oscarbenjamin commented:
> I see that multiple powers of trigonometric functions could be handled much
> more efficiently

I think this only currently works because simplify is being used e.g. the PR now gives:
```python
In [67]: %time print(laplace_transform(sin(t)**6, t, s))
(-I*(-I*(-I*(-I*(-2/((s - 2*I)*((s - 2*I)**2 + 4)) + 2/((s - 4*I)*((s - 4*I)**2 + 4)))/2 + I*(2/((s - 2*I)*((s - 2*I)**2 + 4)) - 2/(s*(s**2 + 4)))/2)/2 + I*(-I*(2/((s - 2*I)*((s - 2*I)**2 + 4)) - 2/(s*(s**2 + 4)))/2 + I*(-2/((s + 2*I)*((s + 2*I)**2 + 4)) + 2/(s*(s**2 + 4)))/2)/2)/2 + I*(-I*(-I*(2/((s - 2*I)*((s - 2*I)**2 + 4)) - 2/(s*(s**2 + 4)))/2 + I*(-2/((s + 2*I)*((s + 2*I)**2 + 4)) + 2/(s*(s**2 + 4)))/2)/2 + I*(-I*(-2/((s + 2*I)*((s + 2*I)**2 + 4)) + 2/(s*(s**2 + 4)))/2 + I*(-2/((s + 4*I)*((s + 4*I)**2 + 4)) + 2/((s + 2*I)*((s + 2*I)**2 + 4)))/2)/2)/2)/2, 0, True)
CPU times: user 185 ms, sys: 1.81 ms, total: 187 ms
Wall time: 186 ms

In [68]: %time print(laplace_transform(sin(t)**6, t, s, simplify=True))
(720/(s*(s**6 + 56*s**4 + 784*s**2 + 2304)), 0, True)
CPU times: user 903 ms, sys: 8.31 ms, total: 912 ms
Wall time: 926 ms
```
So I think removing the call to `simplify` by default probably needs to be accompanied by changing the rules that lead to complicated expressions.

With master we instead have:
```python
In [2]: %time print(laplace_transform(sin(t)**6, t, s))
(720/(s*(s**6 + 56*s**4 + 784*s**2 + 2304)), 0, True)
CPU times: user 1.3 s, sys: 6.81 ms, total: 1.3 s
Wall time: 1.3 s

In [3]: %time print(laplace_transform(sin(t)**6, t, s, simplify=False))
(720/(s*(s**6 + 56*s**4 + 784*s**2 + 2304)), 0, True)
CPU times: user 523 ms, sys: 22 µs, total: 523 ms
Wall time: 521 ms
```
(I guess it just unconditionally simplifies on master)

hanspi42 commented:
You are right, so I will just put the default back to `simplify=True` tomorrow. Then we can identify and improve inefficient rules step by step later in the development.

The convergence plane of your example is indeed wrong; I'll fix that too.

hanspi42 commented:
> I thought this means it should converge for `s > 2` but there is a pole at `s = 4`.

Ouch, there was actually a mistake in two rules. The converge planes were correct, but not the denominators.

hanspi42 committed:
Reverted the default to `simplify=True` and corrected errors in two simple rules

hanspi42 commented:
Changes after simplify-Discussion:

(J) Reverted default to `simplify=True`.  Since we know that we will change
    this eventually, I leave the changes in the test cases; this will also
    help later when we switch it again to `False` after the rules giving
    complicated results have been re-written.
    
(K) Corrected an error in the denominator of the rules for `sinh(a*t)**2`
    and `cosh(a*t)**2` as well as the respective test cases.

hanspi42 committed:
It makes more sense that the `**hints` are passed on to meijerint.

hanspi42 committed:
One test case in test_wester.py changed the output format

oscarbenjamin commented on sympy/integrals/transforms.py:
```diff
@@ -1826,19 +1767,88 @@ def _collapse_extra(self, extra):
+            p.append(res[1])
+            c.append(res[2])
+
+        if _noconds:
+            if _simplify:
+                return Add(*r).simplify(doit=False)
+            else:
+                return Add(*r)
+        else:
+            if _simplify:
+                return Add(*r).simplify(doit=False), Max(*p), And(*c)
+            else:
+                return Add(*r), Max(*p), And(*c)
```
The common logic here can be factored:
```python
result = Add(*r)

if _simplify:
    result = result.simplify(doit=False)

if _noconds:
    plane = Max(*p)
    condition = And(*c)
    return result, plane, condition
else:
    return result
```

hanspi42 replied:
Change (L) below.

oscarbenjamin commented:
It would be useful to have a way to select whether you want to only use the rule-based matching (and skip integration) or only use integration. Then you could compare results from different algorithms.

oscarbenjamin commented on sympy/integrals/transforms.py:
```diff
@@ -1826,19 +1767,88 @@ def _collapse_extra(self, extra):
+                results.append((k*r[0], r[1], r[2]))
+                continue
+            T = None
+            try_directly = not any(func.has(t_)
+                                    for func in f.atoms(AppliedUndef))
+            if try_directly:
```
What exactly is the reason for this check?

hanspi42 replied:
I don't know, it has always been in the `_try_directly` of `IntegralTransform`. All `test_laplace_transform` still run if I remove it, so I do. Change (M)

oscarbenjamin replied:
I'm guessing it's because the integration is unlikely to succeed.

oscarbenjamin commented on sympy/integrals/transforms.py:
```diff
@@ -1826,19 +1767,88 @@ def _collapse_extra(self, extra):
+        T = None
+        try_directly = not any(func.has(self.function_variable)
+                               for func in self.function.atoms(AppliedUndef))
+        if try_directly:
```
Is this a duplication of what is in doit?

hanspi42 replied:
No, it is is a duplicate of the `_try_directly` of the `IntegralTransform` without the lines 

```
        if not fn.is_Add:
            fn = expand_mul(fn)
```

However, my new code does not use its own `_try_directly` method anymore, so I can just as well remove it from the code. Change (N).

oscarbenjamin commented on sympy/integrals/transforms.py:
```diff
@@ -1186,606 +1194,540 @@ def _laplace_build_rules(t, s):
+    as it means nothing to the Laplace transform.
+
+    The function does not remove a factor ``Heaviside(t)``; this is already
+    done by the main function ``laplace_transform``.
```
Is that still true?

hanspi42 replied:
No. Change (O).

oscarbenjamin commented on sympy/integrals/transforms.py:
```diff
@@ -1186,606 +1194,540 @@ def _laplace_build_rules(t, s):
     return laplace_transform_rules
 
-def _laplace_cr(f, a, c, **hints):
+def _laplace_cr(f, **hints):
```
This is not only used by `_compute_transform` so is perhaps not needed any more.

hanspi42 replied:
Removed, change (P).

oscarbenjamin commented on sympy/integrals/transforms.py:
```diff
@@ -1186,606 +1194,540 @@ def _laplace_build_rules(t, s):
-    Laplace transform and returns `None` if it cannot do it.
+def _laplace_ct(f):
+    """
+    Internal helper function that completes the tuple as follows:
```
Given that `_compute_transform` already returns a tuple now is this function needed?

hanspi42 replied:
Yes, I use it in many places to create standard-valued tuples.

oscarbenjamin replied:
The premise of this function is that it tries to hide inconsistent use of types in the code here. It would be better to go to the source and make sure that tuples are already being used everywhere.

hanspi42 replied:
They were used everywhere, so it was easy to remove `_laplace_ct`. Also change (P)

oscarbenjamin commented on sympy/integrals/transforms.py:
```diff
@@ -1860,7 +1870,10 @@ def laplace_transform(f, t, s, legacy_matrix=True, **hints):
+    debug information on by setting ``sympy.SYMPY_DEBUG=True``. The numbers
+    of the rules in the debug information (and the code) refer to Bateman's
+    Tables of Integral Transforms Vol. 1, McGraw-Hill 1954, available here:
+    https://resolver.caltech.edu/CaltechAUTHORS:20140123-101456353
```
We can use a sphinx reference for this:
https://docs.sympy.org/latest/contributing/docstring.html#references-section

hanspi42 replied:
Change (Q)

oscarbenjamin commented:
I've gone through and left comments but otherwise good to merge.

Points to note for the future:

1. Make simplify default to False.
2. Improve formulae for trig rules.
3. Reduce time taken to build rules.

Lastly I suspect that other optimisations are possible and that actually the basic structure used here can be improved. Right now though it is mainly slow because of the three things listed above so those should be fixed before benchmarking and profiling again.

hanspi42 committed:
Last changes: code tidy-up

hanspi42 commented:
Final changes before merge:

(L) Re-factored final step of `doit`.

(M) Removed `trydirectly` check in `doit`.

(N) Removed the now un-used `_try_directly` method; it will again inherit the
    one from `IntegralTransform`.
    
(O) Small doc change re `Heaviside(t)` in `_laplace_rule_heaviside`.

(P) Removed `_laplace_cr` and `_laplace_ct`.

(Q) Moved Bateman54 to a references section in doc.

I think this is it. I will add one more point to your list:

Points for the future:

 - Make simplify default to False.
 - Improve formulae for trig rules.
 - Reduce time taken to build rules.
 - Make it possible to use only rules or only integration.

The next thing I do, though, is the inverse Laplace transform. There I will
already see to it that building the rules needs less time.

oscarbenjamin commented on sympy/integrals/transforms.py:
```diff
@@ -1825,19 +1744,74 @@ def _collapse_extra(self, extra):
+        terms = Add.make_args(fn)
+        results = []
+        for ff in terms:
+            k, f = ff.as_independent(t_, as_Add=False)
+            ft = (f, S.NegativeInfinity, True)
+            d, r = _laplace_apply_simple_rules(ft, t_, s_)
+            if d:
+                results.append((k*r[0], r[1], r[2]))
+                continue
+            d, r = _laplace_apply_prog_rules(ft, t_, s_)
+            if d:
+                results.append((k*r[0], r[1], r[2]))
+                continue
+            d, r = _laplace_expand(ft, t_, s_)
+            if d:
+                results.append((k*r[0], r[1], r[2]))
+                continue
+            T = None
             try:
-                LT = self._compute_transform(fn, t_, s_, **hints)
+                T = self._compute_transform(f, t_, s_, **hints)
             except IntegralTransformError:
-                LT = None
-        return fn, LT
+                T = None
+            if T is not None:
+                results.append((k*T[0], T[1], T[2]))
+            else:
+                results.append((k*LaplaceTransform(f, t_, s_),
+                                S.NegativeInfinity, True))
+
+        r = []
+        p = []
+        c = []
+        for res in results:
+            r.append(res[0])
+            p.append(res[1])
+            c.append(res[2])
```
Looking at this more closely the signature of these rule functions doesn't really make any sense. They all take a tuple `ft = (f, p, c)` as input argument which they immediately unpack. The values of `p` and `c` are always the same on input and although the rule functions use `Max` and `And` this is pointless because e.g. `Max(-oo, pc)` will always just give `pc`. The other place the `p` and `c` arguments are used is when returning `False, (f, p, c)` but there's no point in returning a value along with the False because the value will always be ignored.

I thought that there might be some place where the input plane or convergence conditions would be something other than `-oo, True` but clearly as shown above there is not. There are places that combine convergence conditions with `Max/And` but it happens inside these functions when they call `doit` recursively and so these input values are redundant.

Instead these functions should just take `f` as input and then always return a tuple `(fs, p, c)` or `None` if the rule does not apply. Then this function is just like:
```python
result = _laplace_apply_simple_rules(f, t_, s_)
if result is not None:
    ...
```
Actually this code above is a good candidate for the Walrus operator:
```python
        terms = Add.make_args(fn)
        terms_s = []
        planes = []
        conditions = []

        for term in terms:

            coefficient, ft = term.as_independent(t_, as_Add=False)

            if result := _laplace_apply_simple_rules(ft, t_, s_) is not None):
                pass
            elif result := _laplace_apply_prog_rules(ft, t_, s_) is not None):
                pass
            elif result := _laplace_expand(ft, t_, s_) is not None):
                pass
            elif result := _laplace_integrate(f, t_, s_, **hints) is not None:
                pass
            else:
                f_lt = LaplaceTransform(f, t_, s_)
                result = (f_lt, S.NegativeInfinity, S.true)

            fs, p, c = result
            terms_s.append(coefficient * fs)
            planes.append(p)
            conditions.append(c)
```

hanspi42 replied:
You are right; originally I had plans to be able to use non-ideal planes and conditions as input, but that is, as you say, not necessary anymore. I'll let the Walrus loose ...

hanspi42 committed:
Simplified arguments passed to and from subroutines

hanspi42 commented:
Ultimate final changes before merge:

(R) All rule functions without incoming planes and conditions and return a
    tuple or None.  Code re-written to use the Walrus operator which (believe
    it or not) I did not know existed before you told me abut it.

oscarbenjamin commented:
Why is `LaplaceTransform.doit` being used for recursion instead of `_laplace_transform`?

It looks to me as if `_laplace_transform` is supposed to be the function for actually computing a Laplace transform. Here all of the added rules will not be used when calling `_laplace_transform` though.

Also it seems that `_laplace_transform` has handling of `DiracDelta`. Does that not duplicate what is done in the rule-based part?

The general structure of the code for all of the other transforms is like:
```python
def _hankel_transform(...):
    # Actual code to calculate transform

class HankelTransform(IntegralTransform):
    def _compute_transform(self, ...):
        return _hankel_transform(...)

def hankel_transform(...):
    # End user function for computing transform
    return HankeTransform(...).doit()
```
I would have thought that it should be `_laplace_transform` that would contain all of the code currently in `LaplaceTransform.doit`. and then all of the rules could use `_laplace_transform` for recursion. The existing `_laplace_transform` function could be renamed as e.g. `_laplace_transform_integration` (and changed to return `None` with a signature like the other rules).

Also why specifically does `LaplaceTransform` need to override `doit`? Are there problems with the general `IntegralTransform.doit` function?

hanspi42 committed:
Moved rule logic out of `doit`

hanspi42 commented:
> Also why specifically does `LaplaceTransform` need to override `doit`? Are there problems with the general `IntegralTransform.doit` function?

The `doit` from `IntegralTransform` unfortunately tries integration first and linearity later. But I agree that it is not nice to have everything in `doit`.  So I have now split it up as follows.

Ultra-ultimate final changes before merge:

(R) Renamed `_laplace_transform` to `_laplace_transform_integration`, changed
    its signature, and removed the `DiracDelta` part. It will now simply
    refuse to do anything if it has an incoming `DiracDelta`. Since that
    part was able to solve more complicated arguments to `DiracDelta` by
    integration, I have now modified the code of `_laplace_rule_delta` to
    deal with polynomial-time arguments if all polynomial roots are single.
    It is now even more powerful than before, so there are small changes
    in the tests.
    
    `_laplace_transform_integration` would now try to integrate `f(t)` and
    fail, so the new `_laplace_transform` now also needs to catch that case
    to save time.

(S) Moved all rule stuff to a new `_laplace_transform` and simplified `doit`,
    which now only evaluates `noconds`. 

(T) Replaced all calls to `LaplaceTransform().doit()` in my code to calls to
    the new `_laplace_transform`.  

And now I am sitting here, dreaming of a merge ;)))

hanspi42 committed:
Removed unused import in transforms.py

hanspi42 commented:
By the way, from all tests in `test_laplace_transform`, only three now trigger integration. From stderr:

	Line  145: Integrating exp(2*t**2)*exp(-_s*t) wrt t from 0 to oo.
	Line 1462: Integrating exp(-_s*t)*fresnels(t) wrt t from 0 to oo.
	Line 1539: Integrating exp(-_s*t)*fresnelc(t) wrt t from 0 to oo.

hanspi42 committed:
One bug removed that was only caught by mypy

oscarbenjamin commented:
> The `doit` from `IntegralTransform` unfortunately tries integration first and linearity later.

Okay, then it makes sense to override it.

Are there any situations where linearity breaks convergence?

Some integrals of sums converge even though the integrals if the individual terms do not. I guess for this you would need some functions that grow faster than any exponential but have a difference that does not. Something like:
```python
In [8]: laplace_transform(factorial(t) - t**t, t, s)
Out[8]: 
                  ⎛ t      ⎞                                   
- LaplaceTransform⎝t , t, s⎠ + LaplaceTransform(Γ(t + 1), t, s)
```

oscarbenjamin commented on sympy/integrals/transforms.py:
```diff
@@ -1186,607 +1172,549 @@ def _laplace_build_rules(t, s):
+        elif (r := _laplace_expand(ft, t_, s_)) is not None:
+            pass
+        elif ft.has(AppliedUndef):
+            r = (LaplaceTransform(ft, t_, s_), S.NegativeInfinity, True)
```
Checking only `has(AppliedUndef)` like this will also catch cases where there is a function of a different variable like `laplace_transform(g(x)*sin(t), t, s)`.

hanspi42 replied:
Yes, but such a thing cannot get to that point in the algorithm since it will be treated as a constant by everything else in front of it.

oscarbenjamin replied:
There are many other ways like:
```
In [8]: sin(g(x)*t)
Out[8]: sin(t⋅g(x))
```

hanspi42 replied:
I'll still change it, though

oscarbenjamin commented on sympy/integrals/transforms.py:
```diff
@@ -1186,607 +1172,549 @@ def _laplace_build_rules(t, s):
+        elif (r := _laplace_transform_integration(ft, t_, s_,
+                                      simplify=simplify)) is not None:
+            pass
+        else:
+            r = None
+        if r is not None:
+            results.append((k*r[0], r[1], r[2]))
+        else:
+            results.append((k*LaplaceTransform(ft, t_, s_),
+                            S.NegativeInfinity, True))
+
+    r = []
+    p = []
+    c = []
+    for res in results:
+        r.append(res[0])
+        p.append(res[1])
+        c.append(res[2])
```
Rather than setting `r = None` you could just set
```python
r = (LaplaceTransform(ft, t_, s_), S.NegativeInfinity, True)
```
Then there only needs to be one place to remember to multiply by `k`. We already had a bug from forgetting to multiply by `k` somewhere so it really is good to limit the number of places where we need to keep track of things like that.

Also note that we're packing everything into a list of tuples only to unpack that list again after. We could just directly populate `r, p, c` in the list above like:
```
(ri, pi, ci) = r
r.append(ri)
p.append(pi)
c.append(ci)
```
This also shows that it's confusing to use the same name `r` for two different things and that single letter variable names should generally be discouraged except in particular mathematical contexts.

The code that I wrote above already does this and uses clearer variable names.

hanspi42 replied:
True, good idea. Will change it.

oscarbenjamin commented on sympy/integrals/transforms.py:
```diff
@@ -1186,607 +1172,549 @@ def _laplace_build_rules(t, s):
+        elif ft.has(AppliedUndef):
+            r = (LaplaceTransform(ft, t_, s_), S.NegativeInfinity, True)
+        elif (r := _laplace_transform_integration(ft, t_, s_,
+                                      simplify=simplify)) is not None:
```
To make this a bit clearer above the loop I would perhaps write a little helper function:
```python
def _laplace_integrate(ft, t_, s_):
    # If there are undefined functions f(t) then integration is unlikely to do
    # anything useful so we skip it and given an unevaluated LaplaceTransform.
    has_undefined = any(undef.has(t_) for undef in ft.atoms(AppliedUndef)
    if has_undefined:
        return None
    return _laplace_transform_integrate(ft, t_, s_, simplify=simplify)
        
```

hanspi42 replied:
This I consider overkill, I think my solution that I will submit is clear enough.

oscarbenjamin commented on sympy/integrals/transforms.py:
```diff
@@ -1825,20 +1752,36 @@ def _collapse_extra(self, extra):
+        default behaviour
+        """
+        _noconds = hints.get('noconds', False)
+        _simplify = hints.get('simplify', True)
```
Did we not decide that noconds should default to True here:
```
>>> (2*LaplaceTransform(exp(t), t, s) - 1).doit()
...
AttributeError: 'Tuple' object has no attribute 'as_coeff_Mul'
```

hanspi42 replied:
Yes, we did. I threw it out while refactoring.

oscarbenjamin replied:
Was that intentional?

oscarbenjamin replied:
If not then it shows that a test should be added for this

oscarbenjamin commented:
I promise this is the last round of review and we can get this merged.

sylee957 commented on sympy/integrals/transforms.py:
```diff
@@ -1186,607 +1172,549 @@ def _laplace_build_rules(t, s):
+    expanded term.
+    """
+
+    if f.is_Add:
+        return None
+    r = expand(f, deep=False)
+    if r.is_Add:
+        return _laplace_transform(r, t, s, simplify=False)
+    r = expand_mul(f)
+    if r.is_Add:
+        return _laplace_transform(r, t, s, simplify=False)
+    r = expand(f)
+    if r.is_Add:
+        return _laplace_transform(r, t, s, simplify=False)
+    if not r==f:
+        return _laplace_transform(r, t, s, simplify=False)
+    r = expand(expand_trig(f))
+    if r.is_Add:
+        return _laplace_transform(r, t, s, simplify=False)
+    return None
```
I think that it would be better to separate these things under some utilities like `expand_integral_add, expand_integral_mul, ...` for users to try these stuff manually, 
Such that we don't have to maintain some sort of 'simplify recipe'.

oscarbenjamin replied:
I'm not sure this has such general utility. How would you change this into something that could be used somewhere else?

hanspi42 replied:
This will be part of one of the next steps that we called "Make it possible to use only rules or only integration." in the points for the future.

Those are:
 - Make simplify default to False.
 - Improve formulae for trig rules.
 - Reduce time taken to build rules.
 - Make it possible to use only rules or only integration.

hanspi42 replied:
> Such that we don't have to maintain some sort of 'simplify recipe'.

By the way, that simplify recipe is part of what makes the present algorithm particularly powerful ;)

sylee957 replied:
I mean some general rewriters like `Integral(Add(...), ...)` -> `Add(Integral(...), ...)`

sylee957 replied:
I think that `Integral(f(x) + g(x), x).expand()` works, but doesn't work with laplace transform though

oscarbenjamin replied:
I guess `IntegralTransform` could have an `expand` method to distribute sums and pull out coefficients.

hanspi42 committed:
Last small changed

hanspi42 commented:
> I promise this is the last round of review and we can get this merged.

Hihihi ... did I go overboard with my change-info headers? :)

Ultra-plus ultimate final channges before merge:

(U) `doit` now again has `noconds=True`

(V)  `_laplace_transform` now uses clearer variable names, does not pack/unpack
     unnecessarily, and does the correct `AppliedUndef` check.

hanspi42 commented:
> Some integrals of sums converge even though the integrals if the individual terms do not.

This is so, but the example you give can also not be solved by `integrate(exp(-s*t)*factorial(t) - t**t, (s, 0, oo))`, so I am not really worried by these cases.

oscarbenjamin commented:
> `doit` now again has `noconds=True`

Is there a test for that? The point of adding tests is so these things don't get overlooked when someone in future is trying to make an unrelated change (which is exactly what just happened here).

hanspi42 committed:
Test whether `noconds=True` in `doit`

hanspi42 commented:
> > `doit` now again has `noconds=True`
> 
> Is there a test for that?

Of course ;))

oscarbenjamin commented:
I edited the release notes in the OP. Let me know what you think.

oscarbenjamin commented:
Okay, looks good!

hanspi42 commented:
> I edited the release notes in the OP. Let me know what you think.

What you wrote is very good.

hanspi42 commented:
> Okay, looks good!

Thank you! As always it's a pleasure to work with you; see you soon in another PR.

oscarbenjamin commented:
Yes, thank you also for your work on this.

The main tests have passed so I'm enabling auto-merge.

oscarbenjamin commented:
> thank you also for your work

(and for your patience!)

oscarbenjamin commented:
Now that it's a bit easier to read the debug output I can see some things that are missed by the rule engine:
```python
exp(-t**2-t)
exp(-2*t)*exp(-t**2)
exp(-(t+1)**2)
exp(-t**2-2*t-1)
```
This example is very slow:
```python
cos(10*t-1)
```
Another target for improvement. I think that the rules engine should handle this but it seems to call integrate many times:
```python
cos(t-1)*cosh(t-2)*sin(t-3)*sinh(t-4)*exp(t-5)
```
The debug output has:
```python
In [8]: %time laplace_transform(cos(t-1)*cosh(t-2)*sin(t-3)*sinh(t-4), t, s, simplify=False)
[LT doit] (sin(t - 3)*cos(t - 1)*sinh(t - 4)*cosh(t - 2), t, s)
[LT _l_t] (sin(t - 3)*cos(t - 1)*sinh(t - 4)*cosh(t - 2), t, s)
[LT _l_t] (-sin(1)*sin(t)**2*cos(3)*sinh(2)*sinh(t)**2*cosh(4) + sin(1)*sin(t)**2*cos(3)*sinh(t)*cosh(2)*cosh(4)*cosh(t) + sin(1)*sin(t)**2*cos(3)*sinh(2)*sinh(4)*sinh(t)*cosh(t) - sin(1)*sin(t)**2*cos(3)*sinh(4)*cosh(2)*cosh(t)**2 + sin(1)*sin(3)*sin(t)*cos(t)*sinh(2)*sinh(t)**2*cosh(4) - sin(t)*cos(1)*cos(3)*cos(t)*sinh(2)*sinh(t)**2*cosh(4) + sin(t)*cos(1)*cos(3)*cos(t)*sinh(t)*cosh(2)*cosh(4)*cosh(t) + sin(t)*cos(1)*cos(3)*cos(t)*sinh(2)*sinh(4)*sinh(t)*cosh(t) - sin(1)*sin(3)*sin(t)*cos(t)*sinh(t)*cosh(2)*cosh(4)*cosh(t) - sin(1)*sin(3)*sin(t)*cos(t)*sinh(2)*sinh(4)*sinh(t)*cosh(t) + sin(1)*sin(3)*sin(t)*cos(t)*sinh(4)*cosh(2)*cosh(t)**2 - sin(t)*cos(1)*cos(3)*cos(t)*sinh(4)*cosh(2)*cosh(t)**2 + sin(3)*cos(1)*cos(t)**2*sinh(2)*sinh(t)**2*cosh(4) - sin(3)*cos(1)*cos(t)**2*sinh(t)*cosh(2)*cosh(4)*cosh(t) - sin(3)*cos(1)*cos(t)**2*sinh(2)*sinh(4)*sinh(t)*cosh(t) + sin(3)*cos(1)*cos(t)**2*sinh(4)*cosh(2)*cosh(t)**2, t, s)
[LT _l_t_i ] started with (cos(t)**2*cosh(t)**2, t, _s)
[LT _l_t_i ]     and simplify=False
Integrating exp(-_s*t)*cos(t)**2*cosh(t)**2 wrt t from 0 to oo.
...
```
It seems trig powers don't get handled:
```python
In [21]: %time laplace_transform(sin(t)**2*f(t), t, s, simplify=False)
[LT doit] (f(t)*sin(t)**2, t, s)
[LT _l_t] (f(t)*sin(t)**2, t, s)
CPU times: user 84 ms, sys: 8 ms, total: 92 ms
Wall time: 90 ms
Out[21]: 
⎛                ⎛        2         ⎞          ⎞
⎝LaplaceTransform⎝f(t)⋅sin (t), t, s⎠, -∞, True⎠
```

hanspi42 commented:
> It seems trig powers don't get handled:

Is there a good way to keep a to-do list on github such that you see what I have on my radar and what my priorities are?

oscarbenjamin commented:
You can open an issue and then edit a checklist:

- [X] Handle trig powers
- [ ] Don't recompute rule patterns

hanspi42 commented:
> You can open an issue and then edit a checklist:
> 
>     * [x]  Handle trig powers
> 
>     * [ ]  Don't recompute rule patterns

Done,  #24561.
